src/basic/time-util.c

```diff
@@ -1607,19 +1608,53 @@ bool clock_supported(clockid_t clock) {
 }
 
 int get_timezone(char **ret) {
-        _cleanup_free_ char *t = NULL;
+        _cleanup_free_ char *t = NULL, *dir = NULL, *dir_resolved = NULL, *expected = NULL, *prefix = NULL;
         int r;
 
         assert(ret);
 
-        r = readlink_malloc("/etc/localtime", &t);
+        r = readlink_malloc(etc_localtime(), &t);
         if (r == -ENOENT)
                 /* If the symlink does not exist, assume "UTC", like glibc does */
                 return strdup_to(ret, "UTC");
         if (r < 0)
                 return r; /* Return EINVAL if not a symlink */
 
-        const char *e = PATH_STARTSWITH_SET(t, "/usr/share/zoneinfo/", "../usr/share/zoneinfo/");
+        /* The relative path depends on the depth of etc_localtime().
+         * Note that we cannot resolve the full path since it may contain
+         * symlinks.
+         */
+        r = path_extract_directory(etc_localtime(), &dir);
+        if (r < 0)
+                return r;
+
+        r = chase(dir, NULL, 0, &dir_resolved, NULL);
+        if (r < 0)
+                return r;
+
+        for (const char *cur = dir_resolved;;) {
+                r = path_find_first_component(&cur, /* accept_dot_dot= */ false, NULL);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+                _cleanup_free_ char *new_prefix = NULL;
+                if (prefix) {
+                        new_prefix = path_join(prefix, "..");
+                        freep(&prefix);
+                } else {
+                        new_prefix = strdup("..");
+                }
+                prefix = TAKE_PTR(new_prefix);
```

Comment:
i really don' like/grok this code. in your system, can't you just refuse relative symlinks? i mean, we usually prefer relative ones, but i think it might be ok if we generate absolute ones if the env var is set, to avoid any confision